lib/msal-core/samples/react-sample-app/src/AuthProvider.js

```diff
@@ -0,0 +1,173 @@
+};
+
+const msalApp = new Msal.UserAgentApplication({
+    auth: {
+        clientId: "245e9392-c666-4d51-8f8a-bfd9e55b2456",
+        authority: "https://login.microsoftonline.com/common",
+        validateAuthority: true,
+        postLogoutRedirectUri: "http://localhost:3000"
+    },
+    cache: {
+        cacheLocation: "localStorage",
+        storeAuthStateInCookie: true
+    }
+});
+
+const callMSGraph = async (url, accessToken) => {
+    const response = await fetch(url, {
+        headers: {
+            Authorization: `Bearer ${accessToken}`
+        }
+    });
+
+    return response.json();
+};
+
+const requiresInteraction = errorMessage => {
+    if (!errorMessage || !errorMessage.length) {
+        return false;
+    }
+
+    return (
+        errorMessage.indexOf("consent_required") !== -1 ||
+        errorMessage.indexOf("interaction_required") !== -1 ||
+        errorMessage.indexOf("login_required") !== -1
+    );
+};
+
+const isIE = () => {
+    const ua = window.navigator.userAgent;
+    const msie = ua.indexOf("MSIE ");
+    const msie11 = ua.indexOf("Trident/");
+
+    return msie > 0 || msie11 > 0;
+};
+
+export default C =>
+    class AuthProvider extends Component {
+        static propTypes = {};
+
+        constructor(props) {
+            super(props);
+            this.state = {
+                account: msalApp.getAccount()
+            };
+
+            msalApp.handleRedirectCallback(this.authRedirectCallBack);
+        }
+
+        acquireTokenAndCallMSGraph = async (endpoint, request, redirect) => {
+            const tokenResponse = await msalApp
+                .acquireTokenSilent(request)
+                .catch(error => {
+                    // Call acquireTokenPopup (popup window) in case of acquireTokenSilent failure due to consent or interaction required ONLY
+                    if (requiresInteraction(error.errorCode)) {
+                        return redirect
+                            ? msalApp.acquireTokenRedirect(request)
+                            : msalApp.acquireTokenPopup(request);
+                    }
+                });
+
+            if (tokenResponse) {
+                const graphResponse = await callMSGraph(
+                    endpoint,
+                    tokenResponse.accessToken
+                );
+
+                this.graphAPICallback(graphResponse);
+            }
+        };
+
+        signIn = async () => {
+            if (isIE()) {
+                await msalApp.loginRedirect(loginRequest);
+            } else {
+                await msalApp.loginPopup(loginRequest);
+
+                await this.acquireTokenAndCallMSGraph(
+                    graphConfig.graphMeEndpoint,
+                    tokenRequest
+                );
+
+                this.setState({
+                    account: msalApp.getAccount()
+                });
+            }
+        };
+
+        signOut = () => {
+            msalApp.logout();
+        };
+
+        readMail = async () =>
```

Comment:
I would still encourage breaking this out a little bit more

getAccessTokenForScopes(scopes: Array<String>)

and then the token requests and what not can live in the piece responsible 

```
await const { accessToken } =  this.props.getAccessTokenForScopes(["user.read", "mail.read"]);
await const userMail = this.props.fetchMailFromGraph(accessToken);
this.props.setMailData(userMail);
```

Reply:
maybe I am overcomplicating for a sample
If you think its overkill, maybe just leave a good comment that outlines that 

```js 
// In your application it would be a good idea to generalize the acquisition of tokens away from the fetching of data but for the purpose of this sample we are do them together. 
```